contents/ai/ml.md

```diff
@@ -0,0 +1,149 @@
+
+A machine learning algorithm is also refered to as a model after the algorithm is provided data to learn from.
+
+Prototyping can be broken down into the following phases:  
+1) Basic data preprocessing  
+2) Partitioning of data   
+3) Model training, evaluation and data analysis  
+4) Repeating step 3  
+
+### Basic Data Preprocessing
+Basic data preprocessing is required to remove unwanted noise from the data. Examples of unwanted noise are repeated content and corrupted images. The unwanted content is usually discarded. Training a model on such data hinders its ability to learn the actual patterns in uncorrupted data. 
+
+Statistics of the data is also collected to identify data imbalances. 
+
+![bar_chart](https://upload.wikimedia.org/wikipedia/commons/3/35/Incarceration_Rates_Worldwide_ZP.svg)  
+(image from wikipedia)
+
+Assume that the above barchart indicates the number of data points present for each country in the data. The number of data points for United States is much larger than that of Japan. Class imbalance is present.
+
+Class imbalance is known to negatively impact model performance. One easy remedey is oversampling the classes with fewer instances during training, up till the point where datapoints from each class has the same probability of being picked. This [paper](https://arxiv.org/abs/1710.05381v1) details oversampling and other methods for tackling class imbalance when using Convolutional Neural Networks. The literature review section in the paper also links to techniques effective on other machine learning algorithms.
+
+### Partitioning of Data
+Data is usually split into 3 sets, the training set, validation set and test set.
+
+#### Test Set
+ML models are hardly 100% accurate due to noise in data. Testing the algorithms on real world data is essential for estimating their efficacy. Furthermore, ML models run the risk of overfitting the data they were trained on. Overfitting is a modeling error which occurs when models learn patterns unique to a subset of data. A model which overfits training data performs well on training data but is unable to translate this performance to real world data.
+
+A test set solves the above problem. A test set is created by partitioning the available data. (20-50%, dependent on the total amount of data available) The test set is used only for testing model performance, and will not be touched during model training. The test set should also resemble real world data as much as possible to be a good indicator of real world performance.
+
+When the data is balanced, the test set can be created by randomly sampling the data without replacement. When the data is unbalanced, stratified sampling is performed to ensure all classes of data are adequately represented. This process ensures that no class is skipped during testing, in turn ensuring that test performance is a good indicator of real world performance. 
```

Comment:
You referred to unbalanced data as "data imbalances" just now, it is not very clear if you are referring to the same jargon or not.

Reply:
Good catch there. I think both are used in literature, I'll stick to imbalance as that gives more hits for papers.